types/time.go

```diff
@@ -1615,41 +1615,74 @@ func ExtractDurationNum(d *Duration, unit string) (int64, error) {
 	}
 }
 
-func extractSingleTimeValue(unit string, format string) (int64, int64, int64, float64, error) {
-	fv, err := strconv.ParseFloat(format, 64)
+func parseSingleTimeValue(unit string, format string) (int64, int64, int64, int64, error) {
+	// Format is a preformatted number, it format should be A[.[B]].
+	decimalPointPos := strings.IndexRune(format, '.')
+	if decimalPointPos == -1 {
+		decimalPointPos = len(format)
+	}
+	iv, err := strconv.ParseInt(format[0:decimalPointPos], 10, 64)
 	if err != nil {
 		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
 	}
-	iv := int64(math.Round(fv))
+	riv := iv // Rounded integer value
 
+	dv := int64(0)
+	lf := len(format) - 1
+	// Has fraction part
+	if decimalPointPos < lf {
+		if lf-decimalPointPos >= 6 {
+			// MySQL rounds down to 1e-6.
+			if dv, err = strconv.ParseInt(format[decimalPointPos+1:decimalPointPos+7], 10, 64); err != nil {
```

Comment:
Should we raise a warning for truncated value?

Reply:
MySQL does not raise a warning here, I think it is just a rounding rule instead of truncated value.

Reply:
```sql
MySQL [(none)]> SELECT 10000101000000 + INTERVAL "111111111111111111.499999999999" MICROSECOND;
+-------------------------------------------------------------------------+
| 10000101000000 + INTERVAL "111111111111111111.499999999999" MICROSECOND |
+-------------------------------------------------------------------------+
| 4520-12-21 05:31:51.111111                                              |
+-------------------------------------------------------------------------+
1 row in set, 1 warning (0.001 sec)

MySQL [(none)]> show warnings;
+---------+------+----------------------------------------------------------------------+
| Level   | Code | Message                                                              |
+---------+------+----------------------------------------------------------------------+
| Warning | 1292 | Truncated incorrect INTEGER value: '111111111111111111.499999999999' |
+---------+------+----------------------------------------------------------------------+
1 row in set (0.000 sec)

```

Reply:
Resolved, this error should not add here, in fact, when unit is not SECOND and get a decimal, we need raise a warning.